AppleClang 7.3 reaches maximum recursion depth in pybind11

**Performed steps**
The following steps 

```commandline
pip --version
# pip 19.0.3 from /Users/rlehe/miniconda3/lib/python3.6/site-packages/pip (python 3.6)
cmake --version
# cmake version 3.13.4
cc --version
# Apple LLVM version 7.3.0 (clang-703.0.29)
# Target: x86_64-apple-darwin15.4.0
# Thread model: posix
# InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

pip install openpmd-api
```

led to the following output/problem during install

```
Collecting openpmd-api
  Using cached https://files.pythonhosted.org/packages/fd/82/b41a57fad8a33de8409deb6e47a1106b5f040289b284ab46d2cc3a34c98e/openPMD-api-0.9.0a0.tar.gz
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Requirement already satisfied: numpy<2.0.0,>=1.15.0 in ./miniconda3/lib/python3.6/site-packages (from openpmd-api) (1.16.2)
Building wheels for collected packages: openpmd-api
  Building wheel for openpmd-api (PEP 517) ... error
  Complete output from command /Users/rlehe/miniconda3/bin/python /Users/rlehe/miniconda3/lib/python3.6/site-packages/pip/_vendor/pep517/_in_process.py build_wheel /var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/tmpsmy1xrnq:
  running bdist_wheel
  running build
  running build_ext
  -- The C compiler identification is AppleClang 7.3.0.7030029
  -- The CXX compiler identification is AppleClang 7.3.0.7030029
  -- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
  -- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
  -- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Using the single-header code from /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/share/openPMD/thirdParty/json/single_include/
  -- nlohmann-json: Using INTERNAL version 3.5.0
  -- HDF5: Using hdf5 compiler wrapper to determine C configuration
  -- Found HDF5: /Users/rlehe/miniconda3/lib/libhdf5.dylib;/usr/lib/libpthread.dylib;/Users/rlehe/miniconda3/lib/libz.dylib;/usr/lib/libdl.dylib;/usr/lib/libm.dylib (found suitable version "1.10.4", minimum required is "1.8.13") found components:  C
  -- Can NOT find 'adios_config' - set ADIOS_ROOT, ADIOS_DIR or INSTALL_PREFIX, or check your PATH
  -- Could NOT find ADIOS (missing: ADIOS_LIBRARIES ADIOS_INCLUDE_DIRS) (Required is at least version "1.13.1")
  -- Found PythonInterp: /Users/rlehe/miniconda3/bin/python (found version "3.6.8")
  -- Found PythonLibs: /Users/rlehe/miniconda3/lib/libpython3.6m.dylib
  -- Performing Test HAS_CPP14_FLAG
  -- Performing Test HAS_CPP14_FLAG - Success
  -- pybind11 v2.3.0
  -- pybind11: Using INTERNAL version 2.3.0
  -- MPark.Variant: Using INTERNAL version 1.4.0
  -- Performing Test HAS_FLTO
  -- Performing Test HAS_FLTO - Success
  -- LTO enabled
  -- Note: Skipping Python example tests (missing example files)
  -- Note: run
      . /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/.travis/download_samples.sh
  to add example files to samples/git-sample/ directory!
  
  openPMD build configuration:
    library Version: 0.9.0
    openPMD Standard: 1.1.0
    C++ Compiler: AppleClang 7.3.0.7030029
      /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
  
    Installation prefix: /usr/local
          bin: bin
          lib: lib
      include: include
        cmake: lib/cmake/openPMD
       python: lib/python3.6/site-packages
  
    Additionally, install following third party libraries:
      MPark.Variant: ON
  
    Build Type: Release
    Library: shared
    Testing: OFF
    Invasive Tests: OFF
    Internal VERIFY: ON
    Build Options:
      MPI: OFF
      JSON: ON
      HDF5: ON
      ADIOS1: OFF
      ADIOS2: OFF
      PYTHON: ON
  
  -- Configuring done
  -- Generating done
  -- Build files have been written to: /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/build/temp.macosx-10.7-x86_64-3.6
  Scanning dependencies of target openPMD
  [  4%] Building CXX object CMakeFiles/openPMD.dir/src/Dataset.cpp.o
  [  4%] Building CXX object CMakeFiles/openPMD.dir/src/Datatype.cpp.o
  [  6%] Building CXX object CMakeFiles/openPMD.dir/src/Iteration.cpp.o
  [  8%] Building CXX object CMakeFiles/openPMD.dir/src/IterationEncoding.cpp.o
  [ 10%] Building CXX object CMakeFiles/openPMD.dir/src/Mesh.cpp.o
  [ 12%] Building CXX object CMakeFiles/openPMD.dir/src/ParticlePatches.cpp.o
  [ 14%] Building CXX object CMakeFiles/openPMD.dir/src/ParticleSpecies.cpp.o
  [ 17%] Building CXX object CMakeFiles/openPMD.dir/src/Record.cpp.o
  [ 19%] Building CXX object CMakeFiles/openPMD.dir/src/RecordComponent.cpp.o
  [ 21%] Building CXX object CMakeFiles/openPMD.dir/src/Series.cpp.o
  [ 23%] Building CXX object CMakeFiles/openPMD.dir/src/auxiliary/Filesystem.cpp.o
  [ 25%] Building CXX object CMakeFiles/openPMD.dir/src/backend/Attributable.cpp.o
  [ 27%] Building CXX object CMakeFiles/openPMD.dir/src/backend/BaseRecordComponent.cpp.o
  [ 29%] Building CXX object CMakeFiles/openPMD.dir/src/backend/MeshRecordComponent.cpp.o
  [ 31%] Building CXX object CMakeFiles/openPMD.dir/src/backend/PatchRecord.cpp.o
  [ 34%] Building CXX object CMakeFiles/openPMD.dir/src/backend/PatchRecordComponent.cpp.o
  [ 36%] Building CXX object CMakeFiles/openPMD.dir/src/backend/Writable.cpp.o
  [ 38%] Building CXX object CMakeFiles/openPMD.dir/src/benchmark/mpi/OneDimensionalBlockSlicer.cpp.o
  [ 40%] Building CXX object CMakeFiles/openPMD.dir/src/IO/AbstractIOHandlerHelper.cpp.o
  [ 42%] Building CXX object CMakeFiles/openPMD.dir/src/IO/DummyIOHandler.cpp.o
  [ 44%] Building CXX object CMakeFiles/openPMD.dir/src/IO/IOTask.cpp.o
  [ 46%] Building CXX object CMakeFiles/openPMD.dir/src/IO/HDF5/HDF5IOHandler.cpp.o
  [ 48%] Building CXX object CMakeFiles/openPMD.dir/src/IO/HDF5/ParallelHDF5IOHandler.cpp.o
  [ 51%] Building CXX object CMakeFiles/openPMD.dir/src/IO/JSON/JSONIOHandler.cpp.o
  [ 53%] Building CXX object CMakeFiles/openPMD.dir/src/IO/JSON/JSONIOHandlerImpl.cpp.o
  [ 55%] Building CXX object CMakeFiles/openPMD.dir/src/IO/JSON/JSONFilePosition.cpp.o
  [ 57%] Linking CXX shared library ../lib.macosx-10.7-x86_64-3.6/libopenPMD.dylib
  [ 57%] Built target openPMD
  Scanning dependencies of target openPMD.py
  [ 61%] Building CXX object CMakeFiles/openPMD.py.dir/src/binding/python/AccessType.cpp.o
  [ 61%] Building CXX object CMakeFiles/openPMD.py.dir/src/binding/python/openPMD.cpp.o
  [ 63%] Building CXX object CMakeFiles/openPMD.py.dir/src/binding/python/Attributable.cpp.o
  [ 65%] Building CXX object CMakeFiles/openPMD.py.dir/src/binding/python/BaseRecord.cpp.o
  In file included from /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/src/binding/python/Attributable.cpp:21:
  In file included from /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/share/openPMD/thirdParty/pybind11/include/pybind11/pybind11.h:49:
  In file included from /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/share/openPMD/thirdParty/pybind11/include/pybind11/detail/../attr.h:13:
  In file included from /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/share/openPMD/thirdParty/pybind11/include/pybind11/cast.h:13:
  In file included from /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/share/openPMD/thirdParty/pybind11/include/pybind11/detail/../pytypes.h:12:
  /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/share/openPMD/thirdParty/pybind11/include/pybind11/detail/common.h:461:67: fatal error: recursive template instantiation exceeded maximum depth of 256
  template<size_t N, size_t ...S> struct make_index_sequence_impl : make_index_sequence_impl <N - 1, N - 1, S...> { };
                                                                    ^
  /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/share/openPMD/thirdParty/pybind11/include/pybind11/detail/common.h:461:67: note: in instantiation of template class 'pybind11::detail::make_index_sequence_impl<1, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245>' requested here
  /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/share/openPMD/thirdParty/pybind11/include/pybind11/detail/common.h:461:67: note: in instantiation of template class 'pybind11::detail::make_index_sequence_impl<2, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245>' requested here
  /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/share/openPMD/thirdParty/pybind11/include/pybind11/detail/common.h:461:67: note: in instantiation of template class 'pybind11::detail::make_index_sequence_impl<3, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245>' requested here
  /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/share/openPMD/thirdParty/pybind11/include/pybind11/detail/common.h:461:67: note: in instantiation of template class 'pybind11::detail::make_index_sequence_impl<4, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245>' requested here
  /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/share/openPMD/thirdParty/pybind11/include/pybind11/detail/common.h:461:67: note: in instantiation of template class 'pybind11::detail::make_index_sequence_impl<5, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245>' requested here
  /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/share/openPMD/thirdParty/pybind11/include/pybind11/detail/common.h:461:67: note: (skipping 247 contexts in backtrace; use -ftemplate-backtrace-limit=0 to see all)
  /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/share/openPMD/thirdParty/pybind11/include/pybind11/cast.h:1626:34: note: in instantiation of template class 'std::__1::is_base_of<pybind11::detail::type_caster_generic, pybind11::detail::type_caster<mpark::variant<char, unsigned char, short, int, long, long long, unsigned short, unsigned int, unsigned long, unsigned long long, float, double, long double, std::__1::basic_string<char>, std::__1::vector<char, std::__1::allocator<char> >, std::__1::vector<short, std::__1::allocator<short> >, std::__1::vector<int, std::__1::allocator<int> >, std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long long, std::__1::allocator<long long> >, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >, std::__1::vector<unsigned short, std::__1::allocator<unsigned short> >, std::__1::vector<unsigned int, std::__1::allocator<unsigned int> >, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> >, std::__1::vector<unsigned long long, std::__1::allocator<unsigned long long> >, std::__1::vector<float, std::__1::allocator<float> >, std::__1::vector<double, std::__1::allocator<double> >, std::__1::vector<long double, std::__1::allocator<long double> >, std::__1::vector<std::__1::basic_string<char>, std::__1::allocator<std::__1::basic_string<char> > >, std::__1::array<double, 7>, bool>, void> >' requested here
          detail::enable_if_t<std::is_base_of<type_caster_generic, make_caster<Return>>::value, void>> {
                                   ^
  /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/share/openPMD/thirdParty/pybind11/include/pybind11/pybind11.h:154:42: note: during template argument deduction for class template partial specialization 'return_value_policy_override<type-parameter-0-0, typename enable_if<std::is_base_of<type_caster_generic, make_caster<Return> >::value, void>::type>' [with Return = mpark::variant<char, unsigned char, short, int, long, long long, unsigned short, unsigned int, unsigned long, unsigned long long, float, double, long double, std::__1::basic_string<char>, std::__1::vector<char, std::__1::allocator<char> >, std::__1::vector<short, std::__1::allocator<short> >, std::__1::vector<int, std::__1::allocator<int> >, std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long long, std::__1::allocator<long long> >, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >, std::__1::vector<unsigned short, std::__1::allocator<unsigned short> >, std::__1::vector<unsigned int, std::__1::allocator<unsigned int> >, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> >, std::__1::vector<unsigned long long, std::__1::allocator<unsigned long long> >, std::__1::vector<float, std::__1::allocator<float> >, std::__1::vector<double, std::__1::allocator<double> >, std::__1::vector<long double, std::__1::allocator<long double> >, std::__1::vector<std::__1::basic_string<char>, std::__1::allocator<std::__1::basic_string<char> > >, std::__1::array<double, 7>, bool>]
              return_value_policy policy = return_value_policy_override<Return>::policy(call.func.policy);
                                           ^
  /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/share/openPMD/thirdParty/pybind11/include/pybind11/pybind11.h:72:9: note: in instantiation of function template specialization 'pybind11::cpp_function::initialize<(lambda at /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/src/binding/python/Attributable.cpp:318:31), mpark::variant<char, unsigned char, short, int, long, long long, unsigned short, unsigned int, unsigned long, unsigned long long, float, double, long double, std::__1::basic_string<char>, std::__1::vector<char, std::__1::allocator<char> >, std::__1::vector<short, std::__1::allocator<short> >, std::__1::vector<int, std::__1::allocator<int> >, std::__1::vector<long, std::__1::allocator<long> >, std::__1::vector<long long, std::__1::allocator<long long> >, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >, std::__1::vector<unsigned short, std::__1::allocator<unsigned short> >, std::__1::vector<unsigned int, std::__1::allocator<unsigned int> >, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> >, std::__1::vector<unsigned long long, std::__1::allocator<unsigned long long> >, std::__1::vector<float, std::__1::allocator<float> >, std::__1::vector<double, std::__1::allocator<double> >, std::__1::vector<long double, std::__1::allocator<long double> >, std::__1::vector<std::__1::basic_string<char>, std::__1::allocator<std::__1::basic_string<char> > >, std::__1::array<double, 7>, bool>, openPMD::Attributable &, const std::__1::basic_string<char> &, pybind11::name, pybind11::is_method, pybind11::sibling>' requested here
          initialize(std::forward<Func>(f),
          ^
  /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/share/openPMD/thirdParty/pybind11/include/pybind11/pybind11.h:1112:22: note: in instantiation of function template specialization 'pybind11::cpp_function::cpp_function<(lambda at /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/src/binding/python/Attributable.cpp:318:31), pybind11::name, pybind11::is_method, pybind11::sibling, void>' requested here
          cpp_function cf(method_adaptor<type>(std::forward<Func>(f)), name(name_), is_method(*this),
                       ^
  /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/src/binding/python/Attributable.cpp:318:10: note: in instantiation of function template specialization 'pybind11::class_<openPMD::Attributable>::def<(lambda at /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-install-va0yl636/openpmd-api/src/binding/python/Attributable.cpp:318:31)>' requested here
          .def("get_attribute", []( Attributable & attr, std::string const& key ) {
           ^
  1 error generated.
  make[2]: *** [CMakeFiles/openPMD.py.dir/src/binding/python/Attributable.cpp.o] Error 1
  make[2]: *** Waiting for unfinished jobs....
  make[1]: *** [CMakeFiles/openPMD.py.dir/all] Error 2
  make: *** [all] Error 2
  /private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-build-env-t9dxjzlv/overlay/lib/python3.6/site-packages/setuptools/dist.py:472: UserWarning: Normalizing '0.9.0a' to '0.9.0a0'
    normalized_version,
  Traceback (most recent call last):
    File "/Users/rlehe/miniconda3/lib/python3.6/site-packages/pip/_vendor/pep517/_in_process.py", line 207, in <module>
      main()
    File "/Users/rlehe/miniconda3/lib/python3.6/site-packages/pip/_vendor/pep517/_in_process.py", line 197, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "/Users/rlehe/miniconda3/lib/python3.6/site-packages/pip/_vendor/pep517/_in_process.py", line 141, in build_wheel
      metadata_directory)
    File "/private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-build-env-t9dxjzlv/overlay/lib/python3.6/site-packages/setuptools/build_meta.py", line 192, in build_wheel
      self.run_setup()
    File "/private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-build-env-t9dxjzlv/overlay/lib/python3.6/site-packages/setuptools/build_meta.py", line 234, in run_setup
      self).run_setup(setup_script=setup_script)
    File "/private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-build-env-t9dxjzlv/overlay/lib/python3.6/site-packages/setuptools/build_meta.py", line 141, in run_setup
      exec(compile(code, __file__, 'exec'), locals())
    File "setup.py", line 177, in <module>
      ('License :: OSI Approved :: '
    File "/private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-build-env-t9dxjzlv/overlay/lib/python3.6/site-packages/setuptools/__init__.py", line 145, in setup
      return distutils.core.setup(**attrs)
    File "/Users/rlehe/miniconda3/lib/python3.6/distutils/core.py", line 148, in setup
      dist.run_commands()
    File "/Users/rlehe/miniconda3/lib/python3.6/distutils/dist.py", line 955, in run_commands
      self.run_command(cmd)
    File "/Users/rlehe/miniconda3/lib/python3.6/distutils/dist.py", line 974, in run_command
      cmd_obj.run()
    File "/private/var/folders/4r/rr44bvx10fnb7vsn_158gb7w0000gn/T/pip-build-env-t9dxjzlv/overlay/lib/python3.6/site-packages/wheel/bdist_wheel.py", line 192, in run
      self.run_command('build')
    File "/Users/rlehe/miniconda3/lib/python3.6/distutils/cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "/Users/rlehe/miniconda3/lib/python3.6/distutils/dist.py", line 974, in run_command
      cmd_obj.run()
    File "/Users/rlehe/miniconda3/lib/python3.6/distutils/command/build.py", line 135, in run
      self.run_command(cmd_name)
    File "/Users/rlehe/miniconda3/lib/python3.6/distutils/cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "/Users/rlehe/miniconda3/lib/python3.6/distutils/dist.py", line 974, in run_command
      cmd_obj.run()
    File "setup.py", line 36, in run
      self.build_extension(ext)
    File "setup.py", line 97, in build_extension
      cwd=self.build_temp
    File "/Users/rlehe/miniconda3/lib/python3.6/subprocess.py", line 311, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command '['cmake', '--build', '.', '--config', 'Release', '--', '-j2']' returned non-zero exit status 2.
  
  ----------------------------------------
  Failed building wheel for openpmd-api
  Running setup.py clean for openpmd-api
Failed to build openpmd-api
Could not build wheels for openpmd-api which use PEP 517 and cannot be installed directly
```

This is the pybind11 C++14 maximum template recursion depth issue on older systems described in the FAQ: https://pybind11.readthedocs.io/en/stable/faq.html#recursive-template-instantiation-exceeded-maximum-depth-of-256 We probably have to add `-ftemplate-depth=1024` on those systems.

**System environment**
 - openPMD-api version: 0.9.0a0
 - operating system: `Darwin 15.4.0 Darwin Kernel Version 15.4.0: Fri Feb 26 22:08:05 PST 2016; root:xnu-3248.40.184~3/RELEASE_X86_64 x86_64`
 - machine: Remi's laptop
 - name and version of Python implementation: cpython 3.6.8 from conda

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -23,6 +23,7 @@ Features
 Bug Fixes
 """""""""
 
+- Clang: fix pybind11 compile on older releases, such as AppleClang 7.3-9.0, Clang 3.9 #543
 
 Other
 """""

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -562,6 +562,20 @@ if(openPMD_HAVE_PYTHON)
     )
     target_link_libraries(openPMD.py PRIVATE openPMD)
 
+    # ancient Clang releases
+    #   https://github.com/openPMD/openPMD-api/issues/542
+    #   https://pybind11.readthedocs.io/en/stable/faq.html#recursive-template-instantiation-exceeded-maximum-depth-of-256
+    #   https://bugs.llvm.org/show_bug.cgi?id=18417
+    #   https://github.com/llvm/llvm-project/commit/e55b4737c026ea2e0b44829e4115d208577a67b2
+    if(("${CMAKE_CXX_COMPILER_ID}" STREQUAL "AppleClang" AND
+        CMAKE_CXX_COMPILER_VERSION VERSION_LESS 9.1) OR
+       ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang" AND
+        CMAKE_CXX_COMPILER_VERSION VERSION_LESS 4.0))
+            message(STATUS "Clang: Passing -ftemplate-depth=1024")
+            target_compile_options(openPMD.py
+                PRIVATE -ftemplate-depth=1024)
+    endif()
+
     if(WIN32)
         set(CMAKE_INSTALL_PYTHONDIR_DEFAULT
             "${CMAKE_INSTALL_LIBDIR}\\site-packages"

